src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+﻿using System.Diagnostics.Contracts;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.ServiceModel.Channels
+{
+    internal class ProducerConsumerStream : Stream
```

Comment:
I would recommend reordering the methods to have Write/WriteAsync and Read/ReadAsync one after the other. There are bunch of properties sprinkled in between them currently.